File: datagen/generators/level04.py
"""Level 4 — Two Tones in the Noise. Two on-bin sinusoids buried in unit noise.

Estimator: FFT peak-picking; coherent integration gain ~ N/2 pulls the weak
tone far above the noise floor even though it is invisible in the time domain.
"""

import numpy as np

from datagen.common import field, make_pack, rng, spectrum_of, truth_block

SEED = 1404
N = 2048
DT = 1.0 / 128.0  # fs = 128 Hz, T = 16 s, bin spacing 0.0625 Hz
F1, A1, PHI1 = 12.0, 1.0, 2.1       # k = 192
F2, A2, PHI2 = 33.4375, 0.35, 4.4   # k = 535
SIGMA = 1.0


def build() -> dict:
    g = rng(SEED)
    t = np.arange(N) * DT
    clean = A1 * np.sin(2 * np.pi * F1 * t + PHI1) + A2 * np.sin(2 * np.pi * F2 * t + PHI2)
    y = clean + g.normal(0.0, SIGMA, N)

    bin_spacing = 1.0 / (N * DT)
    se_amp = SIGMA * np.sqrt(2.0 / N)
    return make_pack(
        id=4,
        seed=SEED,
        kind="series",
        columns={"t": t, "y": y},
        spectrum=spectrum_of(y, DT),
        truth=truth_block(
            {"f1": F1, "A1": A1, "f2": F2, "A2": A2, "sigma": SIGMA},
            r"y(t) = 1.0\,\sin(2\pi\cdot 12\,t + \varphi_1) + 0.35\,\sin(2\pi\cdot 33.4375\,t + \varphi_2) + \varepsilon,\ \ \varepsilon\sim\mathcal{N}(0,1)",
            curve_t=t,
            curve_y=clean,
        ),
        grading={
            "mode": "parameter",
            "fields": [
                field("f1", "frequency f₁", F1, bin_spacing / 2, unit="Hz"),
                field("A1", "amplitude A₁", A1, 4 * se_amp),
                field("f2", "frequency f₂", F2, bin_spacing / 2, unit="Hz"),
                field("A2", "amplitude A₂", A2, 4 * se_amp),
            ],
        },
    )

File: src/levels/meta/level04.ts
import type { LevelMeta } from '../../lib/types';

const meta: LevelMeta = {
  id: 4,
  slug: 'two-tones-in-the-noise',
  title: 'Two Tones in the Noise',
  tier: 'FOUNDATIONS',
  tagline: 'one of them you cannot see. find it anyway.',
  brief: String.raw`Sixteen seconds at 128 Hz. Two sinusoids, both drowned in **unit-variance** noise:

$$y(t) = A_1 \sin(2\pi f_1 t + \varphi_1) + A_2 \sin(2\pi f_2 t + \varphi_2) + \varepsilon,\qquad \varepsilon\sim\mathcal{N}(0,1)$$

The second tone has amplitude well below the noise level — invisible in the time domain by construction. Recover $f_1, A_1, f_2, A_2$.

The point of this case: the FFT doesn't just *organize* the data, it *concentrates* it. $N = 2048$ samples of coherent signal pile into one bin while the noise spreads across a thousand.`,
  recover: 'Recover f₁, A₁, f₂, A₂ (both tones are on-bin).',
  starterCode: String.raw`# GROUND TRUTH · CASE 04 — Two Tones in the Noise
t, y = data.t.to_numpy(), data.y.to_numpy()
N = len(y); dt = t[1] - t[0]
print(f"{N} samples at {1/dt:.0f} Hz")
print("time-domain std:", y.std().round(3), "— tone 2 is far below this")

# the spectrum toggle above shows you what to chase. reproduce it,
# quantify it, and mind the units of whatever you read off.
`,
  hints: {
    nudge: String.raw`Same transform as last time — but look at what the noise does in the spectrum vs. what a coherent tone does as $N$ grows.`,
    method: String.raw`Take the FFT, convert to amplitude with $2|X_k|/N$, and pick the two largest peaks (skip the DC bin). Both tones are on-bin, so peak reading is exact for $f$, and $\hat A = 2|X_k|/N$ is unbiased. Sanity-check by overlaying the reconstructed two-tone signal and running \`residuals\` — RMSE should drop to ≈ the noise σ = 1.`,
    derivation: String.raw`Coherent integration: a tone of amplitude $A$ contributes $|X_{k^\*}| = NA/2$ — it grows **linearly** in $N$. White noise contributes to every bin independently with $\mathbb{E}|X_k^{\text{noise}}|^2 = N\sigma^2$, i.e. magnitude $\propto \sqrt{N}$.

In amplitude units ($2|X_k|/N$), the noise floor scales as

$$\sigma_{\text{floor}} = \sigma\sqrt{\frac{2}{N}} = \sqrt{\tfrac{2}{2048}} \approx 0.031$$

so the $A_2 = 0.35$ tone stands $\approx 11$ standard deviations above the floor — trivially detectable in the spectrum, though invisible in the waveform: in the time domain its SNR is $A_2/\sigma = 0.35$, but the FFT gives a processing gain of $\sqrt{N/2} \approx 32\times$.

The maximum of $\sim 1000$ noise bins (Rayleigh-distributed) reaches only $\approx \sigma_{\text{floor}}\sqrt{2\ln 1000} \approx 0.12$, so false peaks stay well below $A_2$. This is how radar, GPS, and gravitational-wave detectors pull signals out of noise that dwarfs them: **watch longer, integrate coherently**.`,
  },
  xLabel: 't (s)',
  yLabel: 'y',
  chart: { spectrumToggle: true },
};

export default meta;
